oauth2/src/main/java/com/okta/spring/oauth/code/OktaOAuthConfig.java

```diff
@@ -0,0 +1,215 @@
+import org.springframework.security.oauth2.client.token.grant.code.AuthorizationCodeResourceDetails;
+import org.springframework.security.oauth2.common.AuthenticationScheme;
+import org.springframework.security.oauth2.config.annotation.web.configuration.OAuth2ClientConfiguration;
+import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+import org.springframework.web.client.RestTemplate;
+
+import javax.annotation.PostConstruct;
+import javax.servlet.Filter;
+import java.util.function.Consumer;
+
+@Configuration
+@Import(OktaOAuthConfig.OktaPropertiesConfiguration.class)
+@ConditionalOnClass({OAuth2ClientConfiguration.class})
+@ConditionalOnBean(OAuth2ClientConfiguration.class)
+@EnableConfigurationProperties(OktaProperties.class)
+public class OktaOAuthConfig extends WebSecurityConfigurerAdapter {
+
+    @Autowired
+    private OAuth2ClientContext oauth2ClientContext;
+
+    @Autowired
+    private OktaProperties oktaProperties;
+
+    @Autowired
+    private PrincipalExtractor principalExtractor;
+
+    @Autowired
+    private AuthoritiesExtractor authoritiesExtractor;
+
+    @Autowired
+    private DiscoveryMetadata discoveryMetadata;
+
+    @Autowired
+    @Qualifier("oktaAuthorizationCodeResourceDetails")
+    private AuthorizationCodeResourceDetails authorizationCodeResourceDetails;
+
+    @Autowired
+    @Qualifier("oktaResourceServerProperties")
+    private ResourceServerProperties resourceServerProperties;
+
+
+    @PostConstruct
+    protected void init() {
+
+        // update properties based on discovery if needed
+
+        updateIfNotSet(oktaProperties.getOauth2()::setIssuer,
+                       oktaProperties.getOauth2().getIssuer(),
+                       discoveryMetadata.getIssuer());
+
+        String issuer = oktaProperties.getOauth2().getIssuer();
+        String baseUrl = issuer.substring(0, issuer.lastIndexOf("/oauth2/"));
+
+        updateIfNotSet(oktaProperties.getClient()::setOrgUrl,
+                       oktaProperties.getClient().getOrgUrl(),
+                       baseUrl);
+    }
+
+    /**
+     * If {code}currentValue{code} is empty, then {code}newValue{code} is applied to {code}setter{code}.
+     * @param setter method to call to update value if needed
+     * @param currentValue the current value to be checked if not empty
+     * @param newValue new value to use if the consumer needs to be called
+     */
+    private void updateIfNotSet(Consumer<String> setter, String currentValue, String newValue) {
+        if (!StringUtils.hasText(currentValue)) {
+            setter.accept(newValue);
+        }
+    }
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+
+        // FIXME: MVC bleed
+        http.authorizeRequests().antMatchers("/okta/*.css").permitAll();
```

Comment:
Spring's fault or ours?

Reply:
Ours, I ran into an ordering issue (?) where a `WebSecurityConfigurerAdapter` from the `thymeleaf` module was not getting applied correctly (and blocking the css).  A little googling showed that I _should_ `apply()` other `WebSecurityConfigurerAdapter`.  Thoughts?